File: 0378-kth-smallest-element-in-a-sorted-matrix/0378-kth-smallest-element-in-a-sorted-matrix.java
class Solution {
    public int kthSmallest(int[][] matrix, int k) {
        int m = matrix.length, n = matrix[0].length;
        //creating a maxheap
        PriorityQueue<Integer> heap = new PriorityQueue<>((a, b) -> Integer.compare(b,a));
        
        for(int i = 0; i < m ; i++){
            for(int j = 0; j < n; j++){
                heap.offer(matrix[i][j]);
                if(heap.size() > k){
                    heap.poll();
                }
            }
        }
        return heap.poll();
    }
}

File: 0378-kth-smallest-element-in-a-sorted-matrix/0378-kth-smallest-element-in-a-sorted-matrix.py
class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        h = []
        for i in range(len(matrix)):
            heapq.heappush(h, [matrix[i][0], i, 0])
        
        while k!=1:
            element, curr_matrix, index = heapq.heappop(h)
            if index + 1 < len(matrix[curr_matrix]):
                heapq.heappush(h, [matrix[curr_matrix][index + 1], curr_matrix, index + 1])
            k -= 1
        
        if h:
            element, curr_matrix, index = heapq.heappop(h)
            return element
        return -1
            